[최용석] Step4 PR 리뷰 요청드립니다.

badaelephant opened the pull request:
안녕하세요 태화님. 

전에 3단계에서 리뷰해주신 부분 반영하여 Step4 리뷰 요청드립니다. 

꼼꼼하게 잘 리뷰해주셔서 감사합니다. ^--^

좋은 하루 되세요

Commit:
refactor: 피드백 반영

1. Distance 원시객체 포장
2. 구간목록(Sections)의 메서드 목적에 맞게 분리
3. 구간목록에서 Optional 적절하게 활용하기 / null 가능성이 없다면 바로 get 하도록 수정
4. Test Fixture 올바르게 사용하기 -> 불필요하게 beforeEach 쓰지말기.
5. 중간 구간 생성에서 추가되는 구간보다 추가할 구간의 거리가 같거나 크면 예외가 발생한다.

Commit:
docs: step4 구간 제거 리드미 작성

Commit:
test: 구간 삭제 인수테스트 작성

Commit:
feat: 구간 삭제 기능 개발

wooobo commented on src/main/java/nextstep/subway/domain/Distance.java:
```diff
@@ -0,0 +1,65 @@
+
+    @Column(name = "distance", nullable = false)
+    private int value;
+    private static final int MIN_DISTANCE = 0;
```
매칭 넘버 상수화 좋네요 👍 
개인적으론 상수는 최상단에 배치하는 편입니다!

wooobo commented on src/main/java/nextstep/subway/domain/Distance.java:
```diff
@@ -0,0 +1,65 @@
+        return this.value == distance.getValue();
+    }
+
+    public int getValue() {
```
`getValue` 위 위치가 `equals`, `hashCode` 사이에 있는게 다소 어색하네요 🤔 

---


```
Integer value= Integer.valueOf("10");
value.intValue();
```
인스턴수 변수가 한개이므로, `distance.value()` 이런 관점으로도 해볼수 있겠네요 😄 
_(수정해야하는 코멘트는 아닙니다!)_

badaelephant replied:
이부분은 제가 정확히 이해하지 못해서 피드백을 반영을 못했는데요, 혹시 말씀해주신대로 변경했을때 어떤 이점이 있을까요??

wooobo commented on src/main/java/nextstep/subway/domain/Section.java:
```diff
@@ -84,18 +82,17 @@ public Station getDownStation() {
-        }
         this.downStation = section.getUpStation();
-        this.distance -= section.getDistance();
+        this.distance = this.distance.subtract(section.getDistance());
```
```suggestion
        this.distance = section.subtractDistance(distance);
```

객체지향적 관점을 고려해보면, `getter`를 줄일수 있겠네요!

badaelephant replied:
오 그렇군요! 수정했습니다.

wooobo commented on src/main/java/nextstep/subway/exception/CannotDeleteException.java:
```diff
@@ -0,0 +1,9 @@
+package nextstep.subway.exception;
+
+public class CannotDeleteException extends RuntimeException {
```
[Why, When and How to Implement Custom Exceptions in Java](https://stackify.com/java-custom-exceptions/)

도움 될까 싶어서,
커스텀 exception에 대한 블로그 공유해요~

badaelephant replied:
불필요한 커스텀 exception은 만들지 않는게 좋겠군요

wooobo commented on src/main/java/nextstep/subway/domain/Sections.java:
```diff
@@ -28,65 +31,90 @@ public void add(Section section) {
+            currentStation = matchingSection(currentStation).getDownStation();
+            orderedStations.add(currentStation);
+        }
+        return orderedStations;
```
`Collections.unmodifiableList` 를 활용해불수 있겠네요 🤔

wooobo commented on src/main/java/nextstep/subway/domain/Sections.java:
```diff
@@ -28,65 +31,90 @@ public void add(Section section) {
+
+    public List<Section> getAll() {
+        return sections;
+    }
+
+
+    public void deleteStation(Station station) throws NotFoundException {
```
```suggestion
    }

    public void deleteStation(Station station) throws NotFoundException {
```

자동 포멧팅 설정해놓으면 편해요~
개행처리가 불규칙해보여요

wooobo commented on src/test/java/nextstep/subway/domain/DistanceTest.java:
```diff
@@ -0,0 +1,44 @@
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.ValueSource;
+
+class DistanceTest {
```
도메인 테스트 👍

wooobo commented on src/test/java/nextstep/subway/controller/SectionAcceptanceTest.java:
```diff
@@ -211,7 +244,70 @@ void createExceptionNoMatch() {
+        ExtractableResponse<Response> 역_조회_결과 = 지하철_역_목록_조회(신분당선ID);
+        SectionResponse 구간_조회_결과 = 지하철_구간_조회(신분당선ID, 상행_구간_ID).as(SectionResponse.class);
+
+        assertAll(
```
```
        // given
        지하철_노선구간_사이에_역_추가_되어_있음(노선ID, 강동구청역ID, SAFE_DISTANCE);

        // when
        ExtractableResponse<Response> 구간제거응답 = 지하철_노선구간_제거_됨(노선ID, 강동구청역ID);

        // then
        지하철_구간_요청_응답_검증(구간제거응답, HttpStatus.NO_CONTENT);
```

다소 많은 검증을 하고 있는것 같기도해요.
제거에 대한 성공 or 실패 여부만 검증하고

`... 거리가 변경된다.` 에 대한 검증은 도메인 테스트로 검증을 해볼 수 도 있을것같아요.
물론, 인수테스트 시나리오에 해당 내역이 포함되어야한다면 큰 문제는 없습니다!

저의 관점 공유차 작성해보았습니다 😄 

---
[9장 단위 테스트](https://catsbi.oopy.io/a4aad7d3-73ed-463d-99aa-c677c9747e44)

wooobo reviewed (changes requested):
안녕하세요 용석님! 
4단계 잘 진행해주셨습니다!
몇가지 코멘트 남겨드렸는데 
확인 후 재요청 부탁드려요~

Commit:
refactor: 기타 마이너 기능 수정

Commit:
refactor: 피드백 반영
1. static field 최상단 위치
2. 불필요한 커스텀 Exception 삭제
3. 자동 포매팅 적용
4. 인수테스트 검증 부분중 도메인에서 테스트 가능한 사항들은 도메인(Sections)으로 이전
5. Optional 용도에 맞게 사용하도록 수정

wooobo commented on src/main/java/nextstep/subway/domain/Distance.java:
```diff
@@ -23,23 +24,30 @@ protected Distance() {
-        }
-    }
 
+    public int value() {
```
`getValue`, `value`  두개를 보았을때,
`Distance.value()`로 충분한 의미가 전달된다고 생각하기에 말씀드린 부분입니다^^

그리고, 이건 지극히 개인적인 부분이지만 객체에게 메시지를 던지는 사고를 하기위한 노력의 과정이기도 해요.
그러한 부분중 하나가 `get-`으로 시작하는 메서드를 만들지 않기위한 노력입니다. 

지극히 개인적인 관점공유였습니다!

wooobo commented on README.md:
```diff
@@ -69,4 +69,21 @@
+1. 기능 요구사항
+    - [x] 종점을 제거하는 경우
+    - [x] 가운데 역을 제거하는 경우
+    - [x] 구간이 하나인 노선에서 역을 제거하는 경우
```
요구사항 잘 정리해주셨네요 :+1:

wooobo reviewed (approved):
피드백에 대해 고민해주셔서 감사합니다 :+1: 

4단계까지 진행하느라 수고하셨습니다!
이만 `merge` 진행하겠습니다!
남은 미션도 응원하겠습니다.
화이팅입니다 :+1:
